src/Feldspar/Compiler/Imperative/Representation.hs

```diff
@@ -337,15 +337,20 @@ renderType VoidType = "void"
 renderType (1 :# t) = renderScalarType t
 renderType (_ :# _) = error "renderType: No support for SIMD vector output"
 renderType StringType = "char *"
-renderType (ArrayType _ t) = renderType t ++ " *"
-renderType (NativeArray _ t) = renderType t
+renderType (ArrayType as _ t) = prependAddressSpace as $ renderType t ++ " *"
+renderType (NativeArray _ _ t) = renderType t
 renderType (StructType n _) = "struct " ++ n
 renderType IVarType{} = "struct ivar"
 
 -- | Extend a helper function for the platform
 extend :: String -> Type -> String
 extend s t = s ++ "_fun_" ++ renderType t
 
+-- | Render a C representation of an address space
+prependAddressSpace :: AddressSpace -> String -> String
```

Comment:
Call this renderAddressSpace, do the prepending in renderType, and put this function before extend.